packages/flutter_tools/lib/src/build_runner/resident_web_runner.dart

```diff
@@ -498,26 +494,21 @@ class _ResidentWebRunner extends ResidentWebRunner {
       }
     }
 
-    Duration transferMarker;
     try {
       if (!deviceIsDebuggable) {
         globals.printStatus('Recompile complete. Page requires refresh.');
-      } else if (!debuggingOptions.buildInfo.isDebug) {
+      } else if (isRunningDebug) {
+        await _vmService.callMethod('hotRestart');
+      } else {
         // On non-debug builds, a hard refresh is required to ensure the
         // up to date sources are loaded.
         await _wipConnection?.sendCommand('Page.reload', <String, Object>{
           'ignoreCache': !debuggingOptions.buildInfo.isDebug,
         });
-      } else {
-        transferMarker = timer.elapsed;
-        await _wipConnection?.debugger?.sendCommand(
-          'Runtime.evaluate', params: <String, Object>{
-            'expression': 'window.\$hotReloadHook([$reloadModules])',
-            'awaitPromise': true,
-            'returnByValue': true,
-          },
-        );
       }
+    } on Exception catch (err) {
+      globals.printError(err.toString());
+      return OperationResult(1, err.toString());
```

Comment:
Will including the `err.toString()` in the `OperationResult` case it to be printed twice?

Reply:
I've just updated here and above to be fatal errors, so we should print the message with a throwToolExit.

I can't think of a case where we hit a HotRestartException or WipError that we could recover from, so getting people out of the tool is likely better than getting stuck in a loop of prompting the user to "Try again after fixing the above error"